### PR TITLE
refactor: use `pathlib` for `file_relative_path`

### PR DESCRIPTION
**Pull request by rexledesma**
## Summary & Motivation
Resolves https://github.com/dagster-io/dagster/issues/21357.

## How I Tested These Changes
existing bk

### CODE REVIEW

**Comment by rexledesma** [comment]
* **#23108** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23108?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rexledesma and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_utils/__init__.py [hunk 1]
```diff
@@ -166,7 +167,7 @@ def file_relative_path(dunderfile: str, relative_path: str) -> str:
     check.str_param(dunderfile, "dunderfile")
     check.str_param(relative_path, "relative_path")
 
-    return os.path.join(os.path.dirname(dunderfile), relative_path)
+    return os.fspath(Path(dunderfile, "..", relative_path).resolve())
```
Comment (sryza) [inline python_modules/dagster/dagster/_utils/__init__.py:170]: is using `.parent` more canonical than `..`?

**Review by sryza — APPROVED** [review]
Nice

File: python_modules/dagster/dagster/_utils/__init__.py (hunk 1 above)
Comment (rexledesma) [reply python_modules/dagster/dagster/_utils/__init__.py:170]: It supposedly is, but it looks incredibly ugly if you're trying to use it multiple times (e.g. `Path("my/path").parent.parent.parent`). So I end up not using it in favor of `..`.

**Pull request merged**